File: macos/Sources/ElevateCore/Models/AccessPackages.swift
import Foundation

/// An entitlement management access package the signed-in user may request.
public struct AccessPackage: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var displayName: String
    public var description: String?
    public var isHidden: Bool

    public init(id: String, displayName: String, description: String? = nil, isHidden: Bool = false) {
        self.id = id
        self.displayName = displayName
        self.description = description
        self.isHidden = isHidden
    }
}

/// Graph's `accessPackageAssignmentRequestState`, plus `unknown` for values this build has not seen.
public enum AccessPackageRequestState: String, Codable, Hashable, Sendable, CaseIterable {
    case submitted, pendingApproval, delivering, delivered, deliveryFailed, denied, scheduled, canceled, partiallyDelivered, unknown

    /// Case-insensitive; anything unrecognised is `unknown` so one new value never fails a page.
    public static func parse(_ raw: String?) -> AccessPackageRequestState {
        guard let raw else { return .unknown }
        return allCases.first { $0.rawValue.caseInsensitiveCompare(raw) == .orderedSame } ?? .unknown
    }

    /// Requested tab: the request has not reached a final state.
    public var isOpen: Bool {
        switch self {
        case .submitted, .pendingApproval, .delivering, .scheduled, .partiallyDelivered: true
        default: false
        }
    }

    /// Declined tab: the request ended without access.
    public var isDeclined: Bool {
        switch self {
        case .denied, .deliveryFailed, .canceled: true
        default: false
        }
    }

    /// Graph accepts a cancel only before delivery starts.
    public var isCancellable: Bool { self == .submitted || self == .pendingApproval }
}

/// One of the signed-in user's own access package requests.
public struct AccessPackageRequest: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var packageId: String
    public var packageName: String
    public var requestType: String
    public var state: AccessPackageRequestState
    /// Graph's free-text status, shown for failed and canceled requests.
    public var status: String?
    public var justification: String?
    public var createdAt: Date?
    public var completedAt: Date?
    public var policyId: String?

    public init(id: String, packageId: String, packageName: String, requestType: String, state: AccessPackageRequestState,
                status: String? = nil, justification: String? = nil, createdAt: Date? = nil, completedAt: Date? = nil, policyId: String? = nil) {
        self.id = id
        self.packageId = packageId
        self.packageName = packageName
        self.requestType = requestType
        self.state = state
        self.status = status
        self.justification = justification
        self.createdAt = createdAt
        self.completedAt = completedAt
        self.policyId = policyId
    }
}

public enum AccessPackageAssignmentState: String, Codable, Hashable, Sendable, CaseIterable {
    case delivering, delivered, expired, unknown

    public static func parse(_ raw: String?) -> AccessPackageAssignmentState {
        guard let raw else { return .unknown }
        return allCases.first { $0.rawValue.caseInsensitiveCompare(raw) == .orderedSame } ?? .unknown
    }
}

/// An access package currently (or formerly) assigned to the signed-in user.
public struct AccessPackageAssignment: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var packageId: String
    public var packageName: String
    public var state: AccessPackageAssignmentState
    public var policyName: String?
    public var expiresAt: Date?

    public init(id: String, packageId: String, packageName: String, state: AccessPackageAssignmentState,
                policyName: String? = nil, expiresAt: Date? = nil) {
        self.id = id
        self.packageId = packageId
        self.packageName = packageName
        self.state = state
        self.policyName = policyName
        self.expiresAt = expiresAt
    }
}

/// One policy the signed-in user may request a package under, from `getApplicablePolicyRequirements`.
public struct PolicyRequirement: Codable, Hashable, Sendable, Identifiable {
    /// The policy id.
    public let id: String
    public var displayName: String
    public var description: String?
    public var isApprovalRequired: Bool
    /// True when the policy asks questions; Elevate hands such requests to the My Access portal.
    public var requiresAnswers: Bool

    public init(id: String, displayName: String, description: String? = nil, isApprovalRequired: Bool, requiresAnswers: Bool) {
        self.id = id
        self.displayName = displayName
        self.description = description
        self.isApprovalRequired = isApprovalRequired
        self.requiresAnswers = requiresAnswers
    }
}

/// What one poll of a tenant returned. Persisted so the next poll can diff against it.
public struct AccessPackageSnapshot: Codable, Hashable, Sendable {
    public var requests: [AccessPackageRequest]
    public var assignments: [AccessPackageAssignment]

    public init(requests: [AccessPackageRequest] = [], assignments: [AccessPackageAssignment] = []) {
        self.requests = requests
        self.assignments = assignments
    }
}

File: windows/src/Elevate.Core/Models/AccessPackages.cs
using System.Text.Json.Serialization;

namespace Elevate.Core.Models;

/// <summary>An entitlement management access package the signed-in user may request.</summary>
public sealed record AccessPackage(string Id, string DisplayName, string? Description = null, bool IsHidden = false);

/// <summary>Graph's <c>accessPackageAssignmentRequestState</c>, plus <see cref="Unknown"/> for values this build has not seen.</summary>
public enum AccessPackageRequestState
{
    Submitted,
    PendingApproval,
    Delivering,
    Delivered,
    DeliveryFailed,
    Denied,
    Scheduled,
    Canceled,
    PartiallyDelivered,
    Unknown,
}

public static class AccessPackageRequestStates
{
    /// <summary>Case-insensitive; anything unrecognised is <see cref="AccessPackageRequestState.Unknown"/> so one new value never fails a page.</summary>
    public static AccessPackageRequestState Parse(string? raw) =>
        ParseName<AccessPackageRequestState>(raw) ?? AccessPackageRequestState.Unknown;

    /// <summary>Requested tab: the request has not reached a final state.</summary>
    public static bool IsOpen(this AccessPackageRequestState state) => state is
        AccessPackageRequestState.Submitted or AccessPackageRequestState.PendingApproval or AccessPackageRequestState.Delivering
        or AccessPackageRequestState.Scheduled or AccessPackageRequestState.PartiallyDelivered;

    /// <summary>Declined tab: the request ended without access.</summary>
    public static bool IsDeclined(this AccessPackageRequestState state) => state is
        AccessPackageRequestState.Denied or AccessPackageRequestState.DeliveryFailed or AccessPackageRequestState.Canceled;

    /// <summary>Graph accepts a cancel only before delivery starts.</summary>
    public static bool IsCancellable(this AccessPackageRequestState state) => state is
        AccessPackageRequestState.Submitted or AccessPackageRequestState.PendingApproval;

    /// <summary>Matches an enum member by name, ignoring case; never by numeric value.</summary>
    internal static T? ParseName<T>(string? raw)
        where T : struct, Enum
    {
        if (string.IsNullOrEmpty(raw))
        {
            return null;
        }

        foreach (var value in Enum.GetValues<T>())
        {
            if (string.Equals(value.ToString(), raw, StringComparison.OrdinalIgnoreCase))
            {
                return value;
            }
        }

        return null;
    }
}

/// <summary>One of the signed-in user's own access package requests.</summary>
public sealed record AccessPackageRequest(
    string Id,
    string PackageId,
    string PackageName,
    string RequestType,
    AccessPackageRequestState State,
    // Graph's free-text status, shown for failed and canceled requests.
    string? Status = null,
    string? Justification = null,
    DateTimeOffset? CreatedAt = null,
    DateTimeOffset? CompletedAt = null,
    string? PolicyId = null);

public enum AccessPackageAssignmentState
{
    Delivering,
    Delivered,
    Expired,
    Unknown,
}

public static class AccessPackageAssignmentStates
{
    public static AccessPackageAssignmentState Parse(string? raw) =>
        AccessPackageRequestStates.ParseName<AccessPackageAssignmentState>(raw) ?? AccessPackageAssignmentState.Unknown;
}

/// <summary>An access package currently (or formerly) assigned to the signed-in user.</summary>
public sealed record AccessPackageAssignment(
    string Id,
    string PackageId,
    string PackageName,
    AccessPackageAssignmentState State,
    string? PolicyName = null,
    DateTimeOffset? ExpiresAt = null);

/// <summary>One policy the signed-in user may request a package under, from <c>getApplicablePolicyRequirements</c>.</summary>
public sealed record PolicyRequirement(
    // The policy id.
    string Id,
    string DisplayName,
    string? Description = null,
    bool IsApprovalRequired = false,
    // True when the policy asks questions; Elevate hands such requests to the My Access portal.
    bool RequiresAnswers = false);

/// <summary>What one poll of a tenant returned. Persisted so the next poll can diff against it.</summary>
public sealed record AccessPackageSnapshot
{
    [JsonConstructor]
    public AccessPackageSnapshot(IReadOnlyList<AccessPackageRequest>? requests = null, IReadOnlyList<AccessPackageAssignment>? assignments = null)
    {
        Requests = requests ?? [];
        Assignments = assignments ?? [];
    }

    public IReadOnlyList<AccessPackageRequest> Requests { get; init; }

    public IReadOnlyList<AccessPackageAssignment> Assignments { get; init; }

    public bool Equals(AccessPackageSnapshot? other) =>
        other is not null && Requests.SequenceEqual(other.Requests) && Assignments.SequenceEqual(other.Assignments);

    public override int GetHashCode() => HashCode.Combine(Requests.Count, Assignments.Count);
}
